golem/managers/payment/pay_all.py

```diff
@@ -30,116 +34,119 @@ def __init__(
+        self._shutdown_timeout = shutdown_timeout.total_seconds()
 
         self._allocation: Optional[Allocation] = None
 
-        self._opened_agreements_count = 0
-        self._closed_agreements_count = 0
-        self._payed_invoices_count = 0
-
         self._event_handlers: List = []
 
+        self._agreements: Dict[str, Agreement] = {}
+        self._agreement_event: asyncio.Event = asyncio.Event()
+
     @trace_span("Starting PayAllPaymentManager", log_level=logging.INFO)
     async def start(self):
-        # TODO: Add stop with event_bus.off()
         self._event_handlers.extend(
             [
                 await self._golem.event_bus.on(NewInvoice, self._pay_invoice_if_received),
                 await self._golem.event_bus.on(NewDebitNote, self._pay_debit_note_if_received),
-                await self._golem.event_bus.on(NewAgreement, self._increment_opened_agreements),
-                await self._golem.event_bus.on(AgreementClosed, self._increment_closed_agreements),
+                await self._golem.event_bus.on(NewAgreement, self._handle_new_agreement),
             ]
         )
 
-    @trace_span("Stopping PayAllPaymentManager", log_level=logging.INFO)
-    async def stop(self):
-        try:
-            await self.wait_for_invoices()
-        except ManagerException:
-            pass
+    @trace_span("Getting allocation", show_results=True, log_level=logging.INFO)
+    async def get_allocation(self) -> Allocation:
+        if self._allocation is None:
+            self._allocation = await self._create_allocation()
 
-        for event_handler in self._event_handlers:
-            await self._golem.event_bus.off(event_handler)
+        return self._allocation
+
+    async def _release_allocation(self) -> None:
+        if self._allocation is None:
+            return
+
+        await self._allocation.release()
 
     @trace_span()
-    async def _create_allocation(self) -> None:
+    async def _create_allocation(self) -> Allocation:
         try:
-            self._allocation = await Allocation.create_any_account(
+            return await Allocation.create_any_account(
                 self._golem, Decimal(self._budget), self._network, self._driver
             )
         except ApiException as e:
             raise ManagerException(json.loads(e.body)["message"]) from e
 
-        # TODO: We should not rely on golem node with cleanups, manager should do it by itself
-        self._golem.add_autoclose_resource(self._allocation)
+    @trace_span("Stopping PayAllPaymentManager", log_level=logging.INFO)
+    async def stop(self):
+        """Terminate all related agreements."""
+        wait_task = asyncio.create_task(asyncio.sleep(self._shutdown_timeout))
+        wait_for_invoices_task = asyncio.create_task(self._wait_for_invoices())
+        _, pending = await asyncio.wait(
+            [wait_task, wait_for_invoices_task], return_when=asyncio.FIRST_COMPLETED
+        )
+        await ensure_cancelled_many(pending)
```

Comment:
Use [asyncio.wait_for](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait_for) as a built-in function for the same flow: calling the coroutine and canceling it after a timeout.